src/Aspire.Hosting.Azure/AzureDeployingContext.cs

```diff
@@ -29,50 +34,358 @@ public async Task DeployModelAsync(AzureEnvironmentResource resource, Cancellati
+                }
+                else
+                {
+                    await resourceTask.CompleteAsync($"Skipped {computeResource.Name} - no deployment target annotation", CompletionState.Completed, cancellationToken).ConfigureAwait(false);
                 }
             }
             catch (Exception ex)
             {
-                var errorMessage = ex switch
+                await resourceTask.CompleteAsync($"Failed to deploy {computeResource.Name}: {ex.Message}", CompletionState.CompletedWithError, cancellationToken).ConfigureAwait(false);
+                throw;
+            }
+        }
+    }
+
+    private static bool TryGetContainerRegistry(IResource computeResource, [NotNullWhen(true)] out IContainerRegistry? containerRegistry)
+    {
+        if (computeResource.TryGetLastAnnotation<DeploymentTargetAnnotation>(out var deploymentTarget) &&
+            deploymentTarget.ContainerRegistry is { } registry)
+        {
+            containerRegistry = registry;
+            return true;
+        }
+
+        containerRegistry = null;
+        return false;
+    }
+
+    private async Task ProcessResourcesForRegistry(IContainerRegistry registry, List<IResource> resources, CancellationToken cancellationToken)
+    {
+        await containerImageBuilder.BuildImagesAsync(
+            resources,
+            new ContainerBuildOptions
+            {
+                TargetPlatform = ContainerTargetPlatform.LinuxAmd64
+            },
+            cancellationToken).ConfigureAwait(false);
+
+        var registryName = await registry.Name.GetValueAsync(cancellationToken).ConfigureAwait(false) ?? throw new InvalidOperationException("Failed to retrieve container registry information.");
+        var acrStep = await activityReporter.CreateStepAsync($"Pushing images to {registryName}", cancellationToken).ConfigureAwait(false);
+        await using (acrStep.ConfigureAwait(false))
+        {
+            try
+            {
+                await AuthenticateToAcr(acrStep, registryName, cancellationToken).ConfigureAwait(false);
+                await PushImageToAcr(acrStep, resources, cancellationToken).ConfigureAwait(false);
+                await acrStep.CompleteAsync($"Successfully pushed {resources.Count} images to {registryName}", CompletionState.Completed, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                await acrStep.CompleteAsync($"Failed to push images to {registryName}: {ex.Message}", CompletionState.CompletedWithError, cancellationToken).ConfigureAwait(false);
+                throw;
+            }
+        }
+    }
+
+    private static void PropagateOutputsToResources(AzureEnvironmentResource resource)
+    {
+        foreach (var (populatedMainOutputName, populatedMainOutputValue) in resource.Outputs)
+        {
+            if (resource.PublishingContext!.ReverseOutputLookup.TryGetValue(populatedMainOutputName, out var outputRef))
```

Comment:
Is there ever a case where we have an output that isn't in the lookup? Should we add an assert here in an `else` case to catch if this ever happens?

Reply:
I'm not sure what a case for this would be. I suspect if this were to happen it would mean that the AzurePublishingContext did something wrong around reconciling the parameters that get referenced and need to be hoisted as outputs? In that case, we probably want to assert that the the state in the publishing context is consistent and throw there? We should never try to deploy if the output lookup is inconsistent.

This might change depending on what we do with the work around splitting up the modules into independent resources.